{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "   ![Image of NCSU](https://brand.ncsu.edu/img/downloads/utilitybar_brick.png)\n",
    "   \n",
    "  \n",
    "      \n",
    "   <font color='Red' size=6>CS590 - Programming and Data Science for Applied Research</font>\n",
    "  \n",
    "   <font color='White'>Space</font>\n",
    "   \n",
    "   <font color='Red' size=6>Course Project</font>                                   \n",
    "    \n",
    "    \n",
    "   <font color='White'>Space</font>\n",
    "   \n",
    "    'SOURCE' (unique)                                                                 \n",
    "    \n",
    "    Description: Simulation to visualize Mendelian sampling variation within Full-Sib Family, explore the IBD Count and IBD Proportion.\n",
    "    \n",
    "    ___AUTHORS___\n",
    "    GS Martins <gsmarti2@ncsu.edu>\n",
    "                                  \n",
    "    Date: DEC 02 2019             \n",
    "                                  \n",
    "    Include: IBD Count, Proportion and Kinship Coefficient calculations\n",
    "             Observed Cotterman's K values\n",
    "             Observed Genetic Relantionship Matrix\n",
    "                                                  \n",
    "            ... diverse -- read comments          \n",
    "                                                  \n",
    "    \n",
    "    Assumptions: Parents A and B are not related ( kinship = 0 ).\n",
    "                 All genes are independent (Do not considers Linkage).\n",
    "                                                                       \n",
    "                                   (12)         (34)                   \n",
    "                                     A --------- B                     \n",
    "                                           |                           \n",
    "                                           |                           \n",
    "                                     C --------- D                     \n",
    "                                                                       \n",
    "                              25% - 13           31 - 25%              \n",
    "                              25% - 14           32 - 25%             \n",
    "                              25% - 23           41 - 25%             \n",
    "                              25% - 24           42 - 25%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#** Defining the size of the random population of i Full-Siblings with j Genes\n",
    "\n",
    "n.sib<-35            #**<<<---- Indicate how many individuals ( suggested < 50 )            \n",
    "n.Gene<-100        #**<<<---- Indicate how many Genes ( suggested < 10000 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\f",
      "\n",
      " \n",
      "Thanks for running this simulation. Let's see the results of our IBD analysis\n",
      " \n",
      "[1] \"The observed 0 IBD allele sharing proportion is 0.248\"\n",
      "[2] \"The observed 1 IBD allele sharing proportion is 0.502\"\n",
      "[3] \"The observed 2 IBD allele sharing proportion is 0.25\" \n",
      "\n",
      "[1] \"The observed average IBD Proportion is 0.50077\"\n",
      "\n",
      "[1] \"The observed min IBD Proportion is 0.41\"\n",
      "[2] \"The observed max IBD Proportion is 0.6\" \n",
      "\n",
      "\n",
      "Remember to check the .csv files saved in your working directory! \n",
      "\n",
      "\n",
      " \n",
      " System time \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "   user  system elapsed \n",
       "   5.50    0.28   10.35 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Press ctrl + enter to 'Source' the script\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# The System Time function will report the time needed to run the entire script.\n",
    "system.time({\n",
    "\n",
    "# Clean the Console\n",
    "cat(\"\\014\") \n",
    "\n",
    "# Limit the number of individuals and genes to save computing time.\n",
    "if(n.Gene>10000\n",
    "   ){\n",
    "  n.Gene=10000\n",
    "}\n",
    "\n",
    "if(n.sib>200\n",
    "   ){\n",
    "  n.sib=200\n",
    "}\n",
    "\n",
    "# POP matrix with Genes on rows, and siblings on columns.\n",
    "# The cell mean corresponds to the genotype of individual i for Gene j\n",
    "\n",
    "# initial pop for vectorization\n",
    "pop<-as.data.frame(matrix(1,1,1))\n",
    "for( i in 1:n.sib ){\n",
    "  for ( j in 1:n.Gene ){\n",
    "  \n",
    "  pop[j,i]<-sample(x = c(\"13\",    # Genotypic classes\n",
    "                         \"14\",\n",
    "                         \"23\",\n",
    "                         \"24\"\n",
    "                         ),\n",
    "                   prob = c(0.25, # Probabilities associated with each class\n",
    "                            0.25,\n",
    "                            0.25,\n",
    "                            0.25),\n",
    "                   size =1,\n",
    "                   replace = T\n",
    "                   )\n",
    "  }\n",
    "}\n",
    "\n",
    "colnames(pop)<-c(1:n.sib)\n",
    "\n",
    "# Creating all possible pairwise combination of n.sib individuals\n",
    "pair<-t(combn(1:n.sib,2))\n",
    "\n",
    "\n",
    "#### Calculating for each one of the possible pairs their average IBD proportion over \n",
    "#    all Genes\n",
    "\n",
    "# Creating vectorized outputs\n",
    "gr<-data.frame()\n",
    "temp<-data.frame()\n",
    "Gene<-character(nrow(pair)*n.sib)\n",
    "Ind.1<-character(nrow(pair)*n.sib)\n",
    "Ind.1.gt<-character(nrow(pair)*n.sib)\n",
    "Ind.2<-character(nrow(pair)*n.sib)\n",
    "Ind.2.gt<-character(nrow(pair)*n.sib)\n",
    "IBD.count<-character(nrow(pair)*n.sib)\n",
    "IBD.prop<-character(nrow(pair)*n.sib)\n",
    "#kin<-character(nrow(pair)*n.sib)\n",
    "all.data<-data.frame()\n",
    "\n",
    "for (i in 1:nrow(pair)){\n",
    "  for (j in 1:n.Gene){\n",
    "    # Counting how many IBD allels for the row 'i' in 'pair'  \n",
    "    temp[j,1]<-length(\n",
    "      intersect(as.vector(strsplit(pop[j,pair[i,1]],\"\"))[[1]],\n",
    "                as.vector(strsplit(pop[j,pair[i,2]],\"\"))[[1]]))\n",
    "  }\n",
    "  # Creating the Genetic Relantionship table (gr)\n",
    "  gr[i,1]<-pair[i,1] # Assigning first indivudal index in column 1\n",
    "  gr[i,2]<-pair[i,2] # Assigning second indivudal index in column 2\n",
    "  gr[i,3]<-( mean(temp[,1]) / 2 ) # Assigning the IBD proportion for Gene j\n",
    "  Gene[((n.Gene*i)-(n.Gene-1)):(n.Gene*i)]<-1:n.Gene\n",
    "  Ind.1[((n.Gene*i)-(n.Gene-1)):(n.Gene*i)]<-pair[i,1]\n",
    "  Ind.1.gt[((n.Gene*i)-(n.Gene-1)):(n.Gene*i)]<-pop[,pair[i,1]]\n",
    "  Ind.2[((n.Gene*i)-(n.Gene-1)):(n.Gene*i)]<-pair[i,2]\n",
    "  Ind.2.gt[((n.Gene*i)-(n.Gene-1)):(n.Gene*i)]<-pop[,pair[i,2]]\n",
    "  IBD.count[((n.Gene*i)-(n.Gene-1)):(n.Gene*i)]<-temp[,1]\n",
    "  IBD.prop[((n.Gene*i)-(n.Gene-1)):(n.Gene*i)]<-temp[,1]/2\n",
    "}\n",
    "\n",
    "all.data<-as.data.frame(cbind(Gene,\n",
    "                              Ind.1,\n",
    "                              Ind.1.gt,\n",
    "                              Ind.2,\n",
    "                              Ind.2.gt,\n",
    "                              IBD.count,\n",
    "                              IBD.prop\n",
    "                              )\n",
    "                        )\n",
    "\n",
    "colnames(all.data)<-c('Gene',\n",
    "                      'Ind.1',\n",
    "                      'Ind.1.gt',\n",
    "                      'Ind.2',\n",
    "                      'Ind.2.gt',\n",
    "                      \"IBD Count\",\n",
    "                      \"IBD Proportion\"\n",
    "                      )\n",
    "\n",
    "colnames(gr)<-c('i',\n",
    "                'j',\n",
    "                \"IBD Proportion\"\n",
    "                )\n",
    "\n",
    "\n",
    "# Creating the Genetic Relantionship Matrix (converting table format from gr to a matrix)\n",
    "grm<-as.data.frame(matrix(1,nrow = n.sib, ncol = n.sib))\n",
    "colnames(grm)<-c(1:n.sib)\n",
    "rownames(grm)<-c(1:n.sib)\n",
    "for (row in 1:n.sib){\n",
    "  for (col in 1:n.sib){\n",
    "    cell.value<-which((gr[,1]==row) & (gr[,2]==col))\n",
    "    if ( row == col ){\n",
    "      grm[row,col]<-1\n",
    "      next\n",
    "    }\n",
    "    else if ( length(cell.value)!=0 ) {\n",
    "      grm[row,col]<-round(gr[cell.value,3],3)\n",
    "    }\n",
    "    else {\n",
    "      cell.value<-which((gr[,1]==col) & (gr[,2]==row))\n",
    "      grm[row,col]<-round(gr[cell.value,3],3)\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "cat(\"\\n\",\"\\n\");flush.console()\n",
    "cat(\"Thanks for running this simulation. Let's see the results of our IBD analysis\")\n",
    "flush.console()\n",
    "cat(\"\\n\",\"\\n\");flush.console()\n",
    "\n",
    "# Observed Cotterman's K\n",
    "print(paste(\"The observed\",c(0:2),\"IBD allele sharing proportion is\",\n",
    "      round(table(all.data$`IBD Count`)/length(all.data$`IBD Count`),3)));flush.console()\n",
    "\n",
    "\n",
    "# Checking the population average IBD proportion and the range \n",
    "# REMEMBER!!! The expected IBD proportion for Full-Sib is 0.5\n",
    "IBD.Proportion<-round(mean(gr[,3]),5)\n",
    "IBD.Range<-round(range(gr[,3]),6)\n",
    "\n",
    "cat(\"\\n\");flush.console()\n",
    "print(paste(\"The observed average IBD Proportion is\",IBD.Proportion));flush.console()\n",
    "cat(\"\\n\");flush.console()\n",
    "print(paste(\"The observed\",c(\"min\",\"max\"),\"IBD Proportion is\",IBD.Range));flush.console()\n",
    "\n",
    "cat(\"\\n\");flush.console()\n",
    "cat(\"\\n\");flush.console()\n",
    "cat(\"Remember to check the .csv files saved in your working directory!\",\"\\n\")\n",
    "flush.console()\n",
    "\n",
    "cat(\"\\n\");flush.console()\n",
    "\n",
    "\n",
    "\n",
    "#EXPORTING THE .CSV FILES\n",
    "\n",
    "# Observed GRM\n",
    "write.csv(grm,paste(\"4_GRMatrix - \",n.sib,\"individuals -\",n.Gene,\" genes.csv\"))\n",
    "\n",
    "# GR table\n",
    "write.csv(gr,paste(\"3_Relationship Table - \",n.sib,\" individuals -\",n.Gene,\" genes.csv\"))\n",
    "\n",
    "# Population genotypic data\n",
    "write.csv(pop,paste(\"1_Population - \",n.sib,\" individuals -\",n.Gene,\" genes.csv\"))\n",
    "\n",
    "# All data\n",
    "write.csv(all.data,paste(\"2_Database - \",n.sib,\" individuals -\",n.Gene,\" genes.csv\"))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Heatmap\n",
    "\n",
    "library(ggplot2)\n",
    "\n",
    "myplot<-ggplot(gr, aes(gr$i, gr$j, fill= gr$`IBD Proportion`))  +\n",
    "ggtitle(paste(\"GRM Heatmap: # Individuals=\",n.sib,\"; # Genes=\",n.Gene)) +\n",
    "xlab(\"Full-sib Individuals\") +\n",
    "ylab(\"Full-sib Individuals\") +\n",
    "theme(legend.title = element_text(size = 30),\n",
    "      legend.text = element_text(size=30),\n",
    "      legend.key.size = unit(1, \"cm\") ,\n",
    "      legend.key.height = unit(0.5, \"cm\"),\n",
    "      plot.title = element_text(size=20, face=\"bold.italic\"),\n",
    "      axis.title.x = element_text(size=10, face=\"bold\"),\n",
    "      axis.title.y = element_text(size=10, face=\"bold\"))+\n",
    "geom_tile() +\n",
    "scale_fill_gradient2(paste(\"IBD Proportion\\n Avg\",\n",
    "                           round(IBD.Proportion,4),\n",
    "                           \"\\n Min \",\n",
    "                           round(min(gr$`IBD Proportion`),4),\n",
    "                           \"\\n Max \",\n",
    "                           round(max(gr$`IBD Proportion`),4)),\n",
    "                     low = \"navyblue\", mid = \"white\", high = \"navyblue\", midpoint = 0.5,\n",
    "                     limits=c(0.39,0.61),\n",
    "                     guide = guide_legend(keywidth = unit(1, \"cm\"))\n",
    "                     )\n",
    "\n",
    "png(paste(getwd(),\"/GRM Heatmap - \",n.sib,\" individuals -\",n.Gene,\" genes.png\",sep=\"\"),\n",
    "    width = 1080,\n",
    "    height = 960)\n",
    "print(myplot)\n",
    "dev.off()\n",
    "graphics.off()\n",
    "\n",
    "cat(\"\\n\",\"\\n\",\"System time\",\"\\n\")\n",
    "print(myplot)\n",
    "rm(list=ls())\n",
    "\n",
    "})#End of system.time function"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
